tools/foreman-cockpit-session

```diff
@@ -0,0 +1,248 @@
+#!/usr/bin/env ruby
+
+require "json"
+require "net/http"
+require "net/ssh"
+
+def dbg(str)
+  # $stderr.write(str + "\n")
+end
+
+# Cockpit protocol, encoding and decoding of control messages.
+
+def send_control(msg)
+  text = JSON.dump(msg)
+  dbg("-> #{text}")
+  $stdout.write("#{text.length+1}\n\n#{text}")
+  $stdout.flush
+end
+
+def read_size(io)
+  size = 0
+  seen = 0
+
+  while seen < 8
+    t = io.readpartial(1)
+
+    unless t
+      return 0
+    end
+
+    if t == "\n"
+      break
+    end
+
+    if t.to_i.zero? && t != "0"
+      raise ArgumentError("Invalid frame: invalid size")
+    end
+
+    size = (size * 10) + t.to_i
+    seen += 1
+  end
+
+  if seen == 8
+    raise ArgumentError("Invalid frame: size too long")
+  end
+
+  size
+end
+
+def read_control()
+  io = IO.for_fd(1)
+  size = read_size(io)
+
+  data = ""
+  while size > 0
+    d = io.readpartial(1)  # XXX - can't we just do readpartial(size)?
+    size -= d.length
+    data += d
+  end
+
+  dbg("<- #{data.strip}")
+  JSON.parse(data)
+end
+
+# Specific control messages
+
+def send_auth_challenge(challenge)
+    send_control({
+        "command" => "authorize",
+        "cookie" => "1234", # must be present, but value doesn't matter
+        "challenge" => challenge
+    })
+end
+
+def send_auth_response(response)
+    send_control({
+        "command" => "authorize",
+        "response" => response
+    })
+end
+
+def read_auth_reply()
+    cmd = read_control()
+    response = cmd["response"]
+    if cmd["command"] != "authorize" || !response
+      raise ArgumentError("Did not receive a valid authorize command")
+    end
+    response
+end
+
+def exit_with_problem(problem, message, auth_methods)
+    send_control({
+        "command" => "init",
+        "problem" => problem,
+        "message" => message,
+        "auth-method-results" => auth_methods
+    })
+    exit 1
+end
+
+# Talking to the Foreman API
+
+def get_token_from_auth_data(auth_data)
+  auth_data.split(" ")[1]
+end
+
+def api_call(path, token)
+  dbg("=> #{path}")
+  uri = URI('http://localhost:3000/api/v2/' + path)
+  req = Net::HTTP::Get.new(uri)
+  req["Cookie"] = "_session_id=#{token}"
+  res = Net::HTTP.start(uri.hostname, uri.port) { |http| http.request(req) }
+  dbg("<= #{res.code} #{res.body[0,500]}")
+  if res.code == "200"
+    return JSON.parse(res.body)
+  elsif res.code == "401"
+    exit_with_problem("authentication-failed",
+                      "Token was not valid",
+                      { "password": "not-tried", "token": "denied" })
+  elsif res.code == "404"
+    return nil
+  else
+    $stderr.write("Error talking to foreman: #{res.body}\n")
+    exit 1
+  end
+end
+
+# SSH via the smart proxy
+
+def ssh_with_proxy(urlstr, params)
```

Comment:
I'd suggest breaking this function into several smaller one so they're easier to read and understand

Reply:
Done.

Now we send and read the headers in a blocking fashion, which should be okay as long as the proxy sends the response header as soon as it gets the request header.  This is currently only the case when then executed program starts writing to something to stdout before reading from stdin.  This is because the proxy currently takes the first byte on stdout as the indication that the program has been sucessfully started. This is the case for cockpit-bridge, which sends a "init" control message immediately, so we should be good in our specific case.